### ISSUE TITLE
cannot execute prepared SELECT more than once

### ISSUE THREAD

**Issue by lexicalunit** [nanodbc/nanodbc#49]
_From @FernandoCorradi on April 23, 2014 18:40_

If I prepare a simple SELECT command and try to execute it more than once, an exception will be thrown. 

According to MS, "To execute a SELECT statement more than once, the application must call SQLCloseCursor before reexecuting the SELECT statement."
http://msdn.microsoft.com/en-us/library/windows/desktop/ms713584%28v=vs.85%29.aspx

Currently there is no way to call SQLCloseCursor using the nanobdc API. I can bypass the problem if I execute a statement::cancel() before the next statement::execute(). Is this the recommended way?

Here is a sample code:

[code]
nanodbc::connection connection("my server connection string");

```
nanodbc::statement statement(connection);
nanodbc::prepare(statement, "SELECT GETDATE() as x");

nanodbc::result rs = statement.execute();
rs.next();
std::cout << rs.get<nanodbc::timestamp>(0) << std::endl;

    // statement.cancel();
    // only by issuing this cancel() the next execute will work
    // but shouldn't we be calling SQLCloseCursor(), as recommended here:
    // http://msdn.microsoft.com/en-us/library/windows/desktop/ms713584%28v=vs.85%29.aspx   

rs = statement.execute();
rs.next();
std::cout << rs.get<nanodbc::timestamp>(0) << std::endl;
```

[/code]

_Copied from original issue: lexicalunit/nanodbc#24_

**Comment by lexicalunit** [comment]
Excellent bug report, thank you for catching this! I just pushed an update that should handle this situation automatically, and added a unit test to check for this going forward.

**Comment by lexicalunit** [comment]
_From @FernandoCorradi on April 24, 2014 14:37_

I liked the way you fixed, but may I suggest a little change? Instead of using SQLCloseCursor, what if we use SQLFreeStmt with SQL_CLOSE option?

As we can see here
http://msdn.microsoft.com/en-us/library/ms709284(v=vs.85).aspx

"Calling SQLFreeStmt with the SQL_CLOSE option is equivalent to calling SQLCloseCursor, except that SQLFreeStmt with SQL_CLOSE does not affect the application if no cursor is open on the statement. If no cursor is open, a call to SQLCloseCursor returns SQLSTATE 24000 (Invalid cursor state)."

Doing so sounds better than using SQLCloseCursor. Since the only thing we check is the executed_ flag, if no cursor was still open, the SQLCloseCursor call will make nanodbc throw an exception. SQLFreeStmt + SQL_CLOSE handles that situation, doing exactly what we need, which is to close any cursor that might be opened, without complaining if there were none.

I am still trying to create a simple sample where the exception I mentioned gets thrown. I get it every time at the same line of code on my application, on the 4th or 5th time I execute that same "SELECT GETDATE()" statement. But there the call is buried inside a DAO class and I also have already executed a lot of stuff (a transaction is opened, selects, inserts and updates). 

Maybe I am doing something wrong in my code and there is no problem at all with the SQLCloseCursor call. But considering the reason we are doing that call in the first place, the SQLFreeStmt(SQL_CLOSE) sounds like a better option. That and the fact that it fixes my application hehe.

I'll try to reproduce the situation in a couple of lines. 

See ya.

**Comment by lexicalunit** [comment]
Great feedback, thanks! Using SQLFreeStmt(SQL_CLOSE) passes tests and seems performant to me; it makes perfect sense to prefer it so as to avoid the case you mentioned above where we could get an invalid cursor state if no cursor is open.

I just pushed a change to use SQLFreeStmt(SQL_CLOSE) as you have suggested. I also replaced executed_ with the existing open_ state variable as that seems to be sufficient and reducing mutable state would be super desirable. Let me know if you have any issues and if you do, I will try to come up with a unit test that reproduces your issue and add the executed_ flag back in :)

Also, I empathize with your frustrations with complexity buried deep within DAO frameworks. In a past life I spent a significant time working with CORBA/TAO and the ACE ORB. Thank you so much for contributing to my little project! I'll re-open this issue pending further feedback.

**Comment by lexicalunit** [comment]
_From @FernandoCorradi on April 24, 2014 20:21_

I believe the one who has to give thanks here is me.

And I finally managed to come up with a short sample that will make nanoodbc throw if it was still using SQLCloseCursor. At lest on SQL Server 2012.

I knew it had something to do with transactions, since I was using one on that code I mentioned, but no matter what I did, the error would not come up when I tried to recreate it. But now I got it. The culprit was the nanodbc::transaction class. It  behaves a little bit different than I was expecting. I like the RAII aproach, but I was thinking that it would rollback transaction if the rollback() method was called before the object was destroyed. Well, after a couple of hours, I finally found out that it does not hehe. It only rolls back when destroyed. That is why my sample was not throwing that same exception the production code was. In the production code the SELECT statement was executed again only after the transaction object was out of scope, while in the sample, since everything was inside the same function, it was still there. Well... nothing well placed new and a delete could not solve :)

Here you go. This will throw with SQLCloseCursor.

```
nanodbc::connection connection("myConnectionString");

nanodbc::statement statement;
statement.prepare(connection, "SELECT GETDATE()");

std::auto_ptr<nanodbc::transaction> transaction(new nanodbc::transaction(connection));

nanodbc::result rs = statement.execute();   
rs.next();
nanodbc::timestamp date = rs.get<nanodbc::timestamp>(0);

transaction.reset(NULL);    

rs = statement.execute();
rs.next();
date = rs.get<nanodbc::timestamp>(0);
```
